### PR TITLE
implement netgame reload level

### PR DESCRIPTION
**Pull request by rfomin**
Should we implement nextlevel netgame command?

### CODE REVIEW

**Comment by fabiangreffrath** [comment]
> Should we implement nextlevel netgame command?

But this won't work when recording a demo, right?

I am a bit afraid about changing the numerical values in `enum buttoncode_t` (who would have expected that?), but with checking for `BT_SPECIAL` and `BTS_RELOAD` at the same time the values become unambigious, right?

**Comment by fabiangreffrath** [comment]
No, wait, how can the game know that this is not meant to be `BT_CHANGE` with `newweapon == 0`?

**Comment by rfomin** [comment]
> > Should we implement nextlevel netgame command?
> 
> But this won't work when recording a demo, right?

Then new demo will be started as with reload. Maybe we should wait for the feature request for this.

> I am a bit afraid about changing the numerical values in `enum buttoncode_t` (who would have expected that?), but with checking for `BT_SPECIAL` and `BTS_RELOAD` at the same time the values become unambiguous, right?

Yes, I think values become unambiguous. In case of connection to Choco/Crispy server, there will be desync, but I think no one would do that. We can check for Choco/Crispy and disable this feature.

> No, wait, how can the game know that this is not meant to be `BT_CHANGE` with `newweapon == 0`?

To be honest, I'm not sure, but it works. This code is convoluted.

**Comment by fabiangreffrath** [comment]
@kraflab Do you think this could collide with any of the changes to `BT_*` handling that you introduced in DSDA?

**Comment by kraflab** [comment]
This won't cause an issue with dsda-doom. The only bit that is interpreted as anything is 1 for pausing. Any new functions only show up in the extended format.

> No, wait, how can the game know that this is not meant to be `BT_CHANGE` with `newweapon == 0`?

The `BT_SPECIAL` bit changes the meaning of all the other bits. So a tick with a pause (128 + 1) is interpreted as a pause and not as an attack (0 + 1).

---

I think you have another problem though. This overloads the 3rd bit for both level reloading and tracking the save slot when saving the game. Saving the game in slot 1 is going to inject `BT_SPECIAL | BTS_SAVEGAME | (1 << BTS_SAVESHIFT)` which happens to equal `BT_SPECIAL | BTS_SAVEGAME | BTS_RELOAD`. You'll need to add guards against misinterpreting this condition.

**Comment by rfomin** [comment]
> I think you have another problem though. This overloads the 3rd bit for both level reloading and tracking the save slot when saving the game. Saving the game in slot 1 is going to inject `BT_SPECIAL | BTS_SAVEGAME | (1 << BTS_SAVESHIFT)` which happens to equal `BT_SPECIAL | BTS_SAVEGAME | BTS_RELOAD`. You'll need to add guards against misinterpreting this condition.

You're right, I'll probably change BTS_RELOAD = 32 to avoid overlapping with BTS_SAVEGAME.

File: Source/g_game.c [hunk 1]
```diff
@@ -660,6 +667,31 @@ static void G_DoLoadLevel(void)
     }
 }
 
+extern int ddt_cheating;
+
+static void G_ReloadLevel(void)
+{
+  int i;
+
+  if (demorecording)
```
Comment (fabiangreffrath) [inline Source/g_game.c:676]: It should still restart from `startmap`, just not when recording a multiplayer demo.

Comment (fabiangreffrath) [reply Source/g_game.c:676]: Oh, and `startepisode` as well. You know, UMAPINFO and stuff. :wink:

**Comment by fabiangreffrath** [comment]
> You're right, I'll probably change BTS_RELOAD = 32 to avoid overlapping with BTS_SAVEGAME.

Alternatively, we could check if `BT_SPECIAL | BTS_RELOAD` and *no other* bit is set.

File: Source/g_game.c (hunk 1 above)
Comment (rfomin) [reply Source/g_game.c:676]: There is no way to know if one of the players is recording a demo in netgame. Alternatively, we can always restart from startmap in netgame.

File: Source/g_game.c [hunk 2]
```diff
@@ -660,6 +667,37 @@ static void G_DoLoadLevel(void)
...
+
+  if (demorecording || netgame)
+  {
+    gamemap = startmap;
+    gameepisode = startepisode;
+  }
+
+  if (demorecording)
+  {
+    ddt_cheating = 0;
```
Comment (rfomin) [inline Source/g_game.c:684]: Why do we need to reset `ddt_cheating` here?

Comment (fabiangreffrath) [reply Source/g_game.c:684]: Because we want to restart the map in a clean non-cheated state.

**Comment by fabiangreffrath** [comment]
Do you consider this ready to merge? Is it free of regressions for single player demo recording?

**Comment by rfomin** [comment]
> Is it free of regressions for single player demo recording?

The difference for a single player is that the reload level is not considered a "new game", so if the user changes the compatibility settings they are not applied. I think this is a minor change.

Another minor issue is that netgame saves have only 3 bits for slot numbers (up to 8).

**Comment by fabiangreffrath** [comment]
> The difference for a single player is that the reload level is not considered a "new game", so if the user changes the compatibility settings they are not applied. I think this is a minor change.

Yes, that's minor. After all it's called "reload level" and not "restart game", so the new behavior is rather expected. 

> Another minor issue is that netgame saves have only 3 bits for slot numbers (up to 8).

But this is the same as before and not related to this PR, right?

**Comment by rfomin** [comment]
> But this is the same as before and not related to this PR, right?

Right. I just didn't know about it until this PR.

**Pull request merged**